### PR TITLE
[RISCV] Register fixed stack slots for callee saved registers for -ms…

### LINKED ISSUE #79491: [RISCV] Post-PEI MIR testing is broken with Zcmp/libcalls
MIR printing is broken when trying to serialize the output after PrologEpilogInserter (`-run-pass=prologepilog` or `-stop-after`).

`llc -mtriple=riscv32 -mattr=+zcmp -x mir -run-pass=prologepilog -verify-machineinstrs -o - `[zcmp-cm-push-pop.mir](https://github.com/llvm/llvm-project/files/14056737/zcmp-cm-push-pop.mir.txt):

```
Assertion failed: (unsigned(ObjectIdx+NumFixedObjects) < Objects.size() && "Invalid Object Idx!"), function isDeadObjectIndex, file MachineFrameInfo.h, line 747.
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
```

Zcmp/save-restore-libcalls have pre-defined frame indices for the saved registers but they're never allocated in `MachineFrameInfo`. When MIRPrinting occurs, the index found in the entries in `MachineFrameInfo::CalleeSavedInfos` is not to be found in `MachineFrameInfo::Objects`, and `isDeadObjectIndex` asserts.

We should create proper (fixed?) slots in MFI in the future to avoid inconsistencies there, since we don't really care about the frame index to be fixed, but the location on the stack.

### PR DESCRIPTION
**Pull request by topperc**
…ave-restore/Zcmp.

PEI previously used fake frame indices for these callee saved registers. These fake frame indices are not register with MachineFrameInfo. This required them to be deleted form CalleeSavedInfo after PEI to avoid breaking later passes. See #79535

Unfortunately, removing the registers from CalleeSavedInfo pessimizes Interprocedural Register Allocation. The RegUsageInfoCollector pass runs after PEI and uses CalleeSavedInfo.

This patch replaces #79535 by properly creating fixed stack objects through MachineFrameInfo. This changes the stack size and offsets returned by MachineFrameInfo which requires changes to how RISCVFrameLowering uses that information.

In addition to the individual object for each register, I've also create a single large fixed object that covers the entire stack area covered by cm.push or the libcalls. cm.push must always push a multiple of 16 bytes and the save restore libcall pushes a multiple of stack align. I think this leaves holes in the stack where we could spill other registers, but it matches what we did previously. Maybe we can optimize this in the future.

The only test changes are due to stack alignment handling after the callee save registers. Since we now have the fixed objects, on the stack the offset is non-zero when an aligned object is processed so the offset gets rounded up, increasing the stack size.

I suspect we might need some more updates for RVV related code. There is very little or maybe even no testing of RVV mixed with Zcmp and save-restore.

### CODE REVIEW

File: llvm/lib/Target/RISCV/RISCVFrameLowering.cpp [hunk 1]
```diff
@@ -1403,6 +1361,93 @@ RISCVFrameLowering::getFirstSPAdjustAmount(const MachineFunction &MF) const {
   return 0;
 }
 
+// Frame indexes representing locations of CSRs which are given a fixed location
+// by save/restore libcalls or Zcmp Push/Pop.
+static const std::pair<unsigned, int> FixedCSRFIMap[] = {
```
Comment (francisvm) [inline llvm/lib/Target/RISCV/RISCVFrameLowering.cpp:1366]:
```suggestion
static const std::pair<Register, int> FixedCSRFIMap[] = {
```

File: llvm/lib/Target/RISCV/RISCVFrameLowering.cpp [hunk 2]
```diff
@@ -291,9 +291,7 @@ static Register getMaxPushPopReg(const MachineFunction &MF,
                                  const std::vector<CalleeSavedInfo> &CSI) {
   Register MaxPushPopReg = RISCV::NoRegister;
   for (auto &CS : CSI) {
-    // RISCVRegisterInfo::hasReservedSpillSlot assigns negative frame indices to
-    // registers which can be saved by Zcmp Push.
-    if (CS.getFrameIdx() < 0)
+    if (llvm::is_contained(AllPopRegs, CS.getReg().id()))
```
Comment (francisvm) [inline llvm/lib/Target/RISCV/RISCVFrameLowering.cpp:294]: Should we merge `AllPopRegs` and `FixedCSRFIMap`?

**Review by francisvm — APPROVED** [review]

File: llvm/lib/Target/RISCV/RISCVFrameLowering.cpp (hunk 1 above)
Comment (topperc) [reply llvm/lib/Target/RISCV/RISCVFrameLowering.cpp:1366]: It should probably be MCPhysReg

**Review by asb — APPROVED** [review]
LGTM.

**Pull request merged**